Bluetooth samples: Add USB Audio to Broadcast Sink sample

fredrikdanebjer opened the pull request:
Added USB Audio output for the Broadcast Sink sample. In addition offloading of the LC3 codec was also made.

fredrikdanebjer commented on samples/bluetooth/broadcast_audio_sink/prj.conf:
```diff
@@ -15,4 +15,8 @@ CONFIG_BT_BUF_ACL_TX_SIZE=251
 CONFIG_BT_DEVICE_NAME="Broadcast Audio Sink"
 
-CONFIG_BT_TINYCRYPT_ECC=y
+# LC3 Needs these
```
I know we should not add these, as they are board specific to whatever board we have that will be using LC3, however I need some input from reviewers on how to solve this properly.

There are two issues:

1. I have added these here since the BSIM `zephyr/tests/bsim/bluetooth/audio_samples/broadcast_audio_sink/tests_scripts/broadcast_audio.sh` will by default compile to `nrf53_bsim`, which today will not pick up the board specific conf file. 

If I do overwrite the current overlay file (which I dont want, I want to append it), it will also:

2. pick the USB component (as it should), however it will then not be able to find the USB device in the device tree. I'm not really sure how to overcome this, as there are currently no examples on how to add a dtsi file when compiling for BSIM.

And and all help would be appreciated!

@aescolar , @Thalley , @MariuszSkamra , @larsgk

larsgk commented on samples/bluetooth/broadcast_audio_sink/boards/nrf5340_audio_dk_nrf5340_cpuapp.conf:
```diff
@@ -1,3 +1,19 @@
+# For LC3 the following configs are needed
+CONFIG_FPU=y
+CONFIG_LIBLC3=y
```
These will now be enabled by Kconfig selects

larsgk commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -43,4 +43,13 @@ config ENABLE_LC3
 	select LIBLC3
 	select FPU
 
+config USE_USB_AUDIO_INPUT
+	bool "Use USB Audio as input"
```
```suggestion
config USE_USB_AUDIO_OUTPUT
	bool "Use USB Audio as output"
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/boards/nrf5340_audio_dk_nrf5340_cpuapp.conf:
```diff
@@ -1,3 +1,19 @@
+CONFIG_BT_ISO_MAX_CHAN=5
+CONFIG_BT_ISO_RX_MTU=60
+#Enable USB
+CONFIG_USE_USB_AUDIO_INPUT=y
```
```suggestion
CONFIG_USE_USB_AUDIO_OUTPUT=y
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,20 +124,101 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+
+	net_buf_unref(ptr_net_buf);
+
+#if defined(CONFIG_USE_USB_AUDIO_INPUT)
```
```suggestion
#if defined(CONFIG_USB_DEVICE_AUDIO)
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,20 +124,101 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+		}
+		return;
+	}
+#endif /*#if defined(CONFIG_USE_USB_AUDIO_INPUT)*/
```
```suggestion
#endif /*#if defined(CONFIG_USB_DEVICE_AUDIO)*/
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+#endif /* defined(CONFIG_LIBLC3) */
 
-static void lc3_decode_handler(struct k_work *work)
+#if defined(CONFIG_USE_USB_AUDIO_INPUT)
```
```suggestion
#if defined(CONFIG_USB_DEVICE_AUDIO)
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+	.data_request_cb = data_request,
+	.data_written_cb = data_written,
+};
+#endif /* defined(CONFIG_USE_USB_AUDIO_INPUT) */
```
```suggestion
#endif /* defined(CONFIG_USB_DEVICE_AUDIO) */
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -274,22 +434,22 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
-#if defined(CONFIG_LIBLC3)
-		int ret;
-		const struct bt_audio_codec_cfg *codec_cfg = &base->subgroups[i].codec_cfg;
+#if defined(CONFIG_ENABLE_LC3)
```
```suggestion
#if defined(CONFIG_LIBLC3)
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -274,22 +434,22 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
+				printk("Error: cannot enable LC3 codec: %d", ret);
+				return;
+			}
+#endif /* defined(CONFIG_ENABLE_LC3) */
```
```suggestion
#endif /* defined(CONFIG_LIBLC3) */
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -696,6 +856,29 @@ static int init(void)
 	}
 
+	/* Initialize ring buffers and USB */
+#if defined(CONFIG_USE_USB_AUDIO_INPUT)
```
```suggestion
#if defined(CONFIG_USB_DEVICE_AUDIO)
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -696,6 +856,29 @@ static int init(void)
+		printk("Failed to enable USB\n");
+		return err;
+	}
+#endif /* defined(CONFIG_USE_USB_AUDIO_INPUT) */
```
```suggestion
#endif /* defined(CONFIG_USB_DEVICE_AUDIO) */
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -862,6 +1045,20 @@ int main(void)
 {
 	int err;
 
+#if defined(CONFIG_USE_USB_AUDIO_INPUT)
```
```suggestion
#if defined(CONFIG_USB_DEVICE_AUDIO)
```

larsgk commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -862,6 +1045,20 @@ int main(void)
+	nrfx_clock_hfclk_start();
+	while (!nrfx_clock_hfclk_is_running()) {
+	}
+#endif /* defined(CONFIG_USE_USB_AUDIO_INPUT) */
```
```suggestion
#endif /* defined(CONFIG_USB_DEVICE_AUDIO) */
```

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -30,6 +40,30 @@ BUILD_ASSERT(IS_ENABLED(CONFIG_SCAN_SELF) || IS_ENABLED(CONFIG_SCAN_OFFLOAD),
+#define USB_SAMPLE_RATE	             (48000U)
+#define USB_FRAME_DURATION_US        (1000U)
+#define USB_TX_BUF_NUM               (10U)
+#define BROADCAST_MONO_SAMPLE_SIZE   (MAX_NUM_SAMPLES_MONO * sizeof(int16_t))
+#define BROADCAST_STEREO_SAMPLE_SIZE (BROADCAST_MONO_SAMPLE_SIZE * sizeof(int16_t))
+#define USB_STEREO_SAMPLE_SIZE       ((USB_FRAME_DURATION_US * USB_SAMPLE_RATE * sizeof(int16_t) \
+				      * 2) / USEC_PER_SEC)
```
The `sizeof(int16_t)` is because the data is int16_t? 

Adding a comment or a new macro to denote that might be useful

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
-static void lc3_decode_handler(struct k_work *work)
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+/* Duplicate the audio from one channel and put it in both channels */
+void mix_mono_to_stereo(void)
```
```suggestion
static void mix_mono_to_stereo(void)
```

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
-	int octets_per_frame;
-	struct broadcast_sink_stream *sink_stream = CONTAINER_OF(
-		k_work_delayable_from_work(work), struct broadcast_sink_stream, lc3_decode_work);
+	int16_t usb_audio_data[MAX_NUM_SAMPLES_MONO] = {0};
+	int16_t usb_audio_data_stereo[MAX_NUM_SAMPLES_STEREO] = {0};
```
How large are these? Does it make sense to stack allocate them?

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+	size = ring_buf_get(&streams[0].audio_ring_buf, (uint8_t *)usb_audio_data,
+			    MAX_NUM_SAMPLES_STEREO);
+	if (size != MAX_NUM_SAMPLES_STEREO) {
+		memset(&((uint8_t *)usb_audio_data)[size], 0, MAX_NUM_SAMPLES_STEREO - size);
```
With `MAX_NUM_SAMPLES_STEREO - size` you are mixing a count and a size which doesn't sound right

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+	size = ring_buf_get(&ring_buf_usb, (uint8_t *)usb_audio_data, USB_STEREO_SAMPLE_SIZE);
+	if (size != USB_STEREO_SAMPLE_SIZE) {
+		memset(&((uint8_t *)usb_audio_data)[size], 0, USB_STEREO_SAMPLE_SIZE);
+	}
+	pcm_buf = net_buf_alloc(&usb_tx_buf_pool, K_NO_WAIT);
```
```suggestion
	}
	
	pcm_buf = net_buf_alloc(&usb_tx_buf_pool, K_NO_WAIT);
```

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -122,60 +244,101 @@ static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
-			printk("  decoder failed - wrong parameters?\n");
-		}
+	out = net_buf_add(pcm_buf, USB_STEREO_SAMPLE_SIZE);
+	memset(out, 0, USB_STEREO_SAMPLE_SIZE);
+	memcpy(out, usb_audio_data, USB_STEREO_SAMPLE_SIZE);
```
Why do we need the `memset` here? Aren't we writing the same size to the same pointer twice?

fredrikdanebjer replied:
Yes, this is a mistake

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -274,22 +434,22 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
-		}
+			if (codec_cfg->id != BT_HCI_CODING_FORMAT_LC3) {
+				printk("Unsupported codec 0x%02x", codec_cfg->id);
+				return;
+			}
 
-		ret = lc3_enable(codec_cfg);
-		if (ret < 0) {
-			printk("Error: cannot enable LC3 codec: %d", ret);
-			return;
-		}
+			ret = lc3_enable(codec_cfg, &streams[j]);
+			if (ret < 0) {
+				printk("Error: cannot enable LC3 codec: %d", ret);
+				return;
+			}
```
This won't work if `base->subgroup_count > 1` as you will be initializing `streams[0]` more than once

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -696,6 +856,29 @@ static int init(void)
+	int ret;
+	const struct device *hs_dev = DEVICE_DT_GET(DT_NODELABEL(hs_0));
+
+	for (size_t i = 0U; i < CONFIG_BT_BAP_BROADCAST_SNK_STREAM_COUNT; i++) {
```
```suggestion
	for (int i = 0; i < CONFIG_BT_BAP_BROADCAST_SNK_STREAM_COUNT; i++) {
```

Kconfig values are `int`s

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -862,6 +1045,20 @@ int main(void)
 {
 	int err;
 
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+	/* Use this to turn on 128 MHz clock for cpu_app, this improves the  */
+	err = nrfx_clock_divider_set(NRF_CLOCK_DOMAIN_HFCLK, NRF_CLOCK_HFCLK_DIV_1);
+
+	err -= NRFX_ERROR_BASE_NUM;
+	if (err) {
+		return err;
+	}
+
+	nrfx_clock_hfclk_start();
+	while (!nrfx_clock_hfclk_is_running()) {
+	}
+#endif /* defined(CONFIG_USB_DEVICE_AUDIO) */
```
Is this necessary?

fredrikdanebjer replied:
It's a very good question, and please feel free to involve anyone else from Nordic understanding this better.

I see (or hear) much better result when I do include this, esp. on 24Khz. 16Khz sounds good with or without, but 24Khz is really struggling, like you are listening to radio in your car and entering a tunnel without it.
There are still a few "sparks" left even with this, but its a huge improvement.

We did see similar results on source previously as well, but IIRC @larsgk said that that solved itself once the threads were offloaded. But it doesn't seem to have helped here.

Thalley replied:
Have you tried running the thread analyzer to see if we are at 100% CPU usage? From what I've seen before, decoding a single 16KHz stream only uses something like 17% usage without this overclocking, so I'm surprised you are seeing it. 

If you are hearing issue with the audio and the CPU is not at 100%, then the issue may lie elsewhere and should probably be figured out rather than just pumping up the CPU

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,20 +124,101 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+{
+	while (true) {
+		k_sem_take(&lc3_decoder_sem, K_FOREVER);
+		/* For now we only handle one channel, so always only decode the first */
+		do_lc3_decode(&streams[0]);
+
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+		/* If the ring buffer usage is alrger than zero, then there is data to process */
+		if (ring_buf_space_get(&streams[0].audio_ring_buf)) {
+			mix_mono_to_stereo();
```
You claim that we only use 1 channel, but there's nothing in the code below that ensures that.

fredrikdanebjer replied:
That is correct, this code is the actual guarantee, we will only take data from one channel.

I do want to extend to handle stereo here properly, but for now I would want to keep this PR utilizing only one channel.

Thalley replied:
Right. 

Whether to mix mono to stereo will also depend on what the stream we are synced to does. 

This assumes that mono will always be sent in the stream without checking the codec configuration of that channel, so that should also be ensure before mixing mono, right?

A single stream (BIS) can send stereo as well

Thalley replied:
> I also understand that you could have stereo on a single BIS, and that is basically what I assume right? That the audio is mixed as stereo on a single channel, and then it's duplicated for both USB channels.

Not fully sure. 

If you do stereo on a single BIS, then the left and right channels come after each other (at least for LC3, as it is a mono codec). So if the broadcast source is sending stereo on a single BIS with frame size = 40, then the SDU size is 2 x 40 => 80. 

If you are not doing proper stereo handling in this PR, then I suggest to remove all of it. Having a half-implementation of something in a sample isn't a good solution IMO.

So either only support mono, or properly support stereo.

fredrikdanebjer commented on samples/bluetooth/broadcast_audio_sink/boards/nrf5340_audio_dk_nrf5340_cpuapp.conf:
```diff
@@ -1,3 +1,16 @@
+CONFIG_BT_ISO_RX_BUF_COUNT=20
+CONFIG_BT_ISO_SYNC_RECEIVER=y
+CONFIG_BT_ISO_MAX_CHAN=5
+CONFIG_BT_ISO_RX_MTU=60
```
The reason why its here is because its only needed when we enable LC3 and USB, which we only do for nrf53

Thalley replied:
The `CONFIG_BT_ISO_RX_MTU` has nothing to do with LC3. 

It's a max value, so setting it to 310 is a safe bet, but we may want to use a lower value if we only want to support the mandatory configuration which is indeed 60 (but that goes for the sample itself, and not related to LC3).

larsgk commented:
I removed the `stale` label as I know @fredrikdanebjer will look at this PR now

fredrikdanebjer committed:
Bluetooth samples: Add USB Audio to Broadcast Sink sample

Added USB Audio output for the Broadcast Sink sample. In addition
offloading of the LC3 codec was also made. The sample supports only
mono, and a KConfig option was added to configure which audio
location to sync to.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

fredrikdanebjer commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,20 +124,101 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+		err = lc3_decode(sink_stream->lc3_decoder, buf_data + offset, octets_per_frame,
+						 LC3_PCM_FORMAT_S16, audio_buf, 1);
+
+		if (err == 1) {
+			printk("  decoder performed PLC\n");
+		} else if (err < 0) {
+			printk("  decoder failed - wrong parameters? (err = %d)\n", err);
+		}
```
Then we have an ordinary success without PLC, this would be the normal case (and would flood the output if printed)

fredrikdanebjer commented on samples/bluetooth/broadcast_audio_sink/boards/nrf5340_audio_dk_nrf5340_cpuapp.conf:
```diff
@@ -1,3 +1,16 @@
+CONFIG_BT_ISO_MAX_CHAN=5
+CONFIG_BT_ISO_RX_MTU=60
+#Enable USB
+CONFIG_USE_USB_AUDIO_OUTPUT=y
```
`CONFIG_USE_USB_AUDIO_OUTPUT` will select `CONFIG_USB_DEVICE_AUDIO`, its in the sample Kconfig

Thalley commented on samples/bluetooth/broadcast_audio_sink/boards/nrf5340_audio_dk_nrf5340_cpuapp.conf:
```diff
@@ -1,3 +1,16 @@
+CONFIG_BT_ISO_RX_MTU=60
+#Enable USB
+CONFIG_USE_USB_AUDIO_OUTPUT=y
+CONFIG_USB_DEVICE_PRODUCT="USB Broadcast Sink"
```
> Not sure why I would get a warning?

Because you are setting a value that depends on another Kconfig that is optional. 

For example, if I build with `CONFIG_ENABLE_LC3=n` then I get the following warnings:
```
warning: USE_USB_AUDIO_OUTPUT (defined at
/home/thalley/dev/nordic/zephyrproject/zephyr/samples/bluetooth/broadcast_audio_sink/Kconfig:46) was
assigned the value 'y' but got the value 'n'. Check these unsatisfied dependencies: ENABLE_LC3 (=n).
See http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_USE_USB_AUDIO_OUTPUT and/or look up
USE_USB_AUDIO_OUTPUT in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.


warning: USB_DEVICE_PRODUCT (defined at subsys/usb/device/Kconfig:37) was assigned the value 'USB
Broadcast Sink' but got the value ''. Check these unsatisfied dependencies: USB_DEVICE_STACK (=n).
See http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_USB_DEVICE_PRODUCT and/or look up
USB_DEVICE_PRODUCT in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.
```

That might be OK since it's enabled by default though

fredrikdanebjer replied:
I think I will add `CONFIG_ENABLE_LC3=y` in this conf file as well to make this clearer.

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,99 +122,221 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
 
-#if defined(CONFIG_LIBLC3)
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+int16_t usb_audio_data[MAX_NUM_SAMPLES_MONO] = {0};
+int16_t usb_audio_data_stereo[MAX_NUM_SAMPLES_STEREO] = {0};
```
```suggestion
static int16_t usb_audio_data[MAX_NUM_SAMPLES_MONO] = {0};
static int16_t usb_audio_data_stereo[MAX_NUM_SAMPLES_STEREO] = {0};
```

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,99 +122,221 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+	}
+
+	/* Put in ring-buffer to be consumed */
+	rbret = ring_buf_put(&sink_stream->audio_ring_buf, (uint8_t *)audio_buf,
+				      BROADCAST_MONO_SAMPLE_SIZE);
```
Some indentation is off here

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,99 +122,221 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+	uint32_t rbret;
+
+	if (ring_buf_space_get(&sink_stream->audio_ring_buf) == 0) {
+		ring_buf_reset(&sink_stream->audio_ring_buf);
+	}
```
Can you elaborate (in a code comment perhaps?) why we reset the buffer when there is no more space left?

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,99 +122,221 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
-	lc3_decoder = lc3_setup_decoder(frame_duration_us, freq_hz, 0, /* No resampling */
-					&lc3_decoder_mem);
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+	sink_stream->lc3_decoder = lc3_setup_decoder(frame_duration_us, freq_hz, USB_SAMPLE_RATE,
+					&sink_stream->lc3_decoder_mem);
+#else
+	sink_stream->lc3_decoder = lc3_setup_decoder(frame_duration_us, freq_hz, 0,
+					&sink_stream->lc3_decoder_mem);
```
Indentation

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,99 +122,221 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+	if (size != BROADCAST_STEREO_SAMPLE_SIZE) {
+		static int rb_put_failures;
 
-	buf_data = ptr_net_buf->data;
-	octets_per_frame = ptr_net_buf->len / frames_per_sdu;
+		rb_put_failures++;
+		if (rb_put_failures % 1000 == 0) {
+			printk("%s: Failure to add to ring buffer %d, %u\n", __func__,
+			       rb_put_failures, size);
+		}
```
We never need to reset the `rb_put_failures`?

Thalley reviewed (changes requested):
A few minor comments, so aside from the stereo/mono things, it looks pretty good

larsgk commented:
I can add that I tried this draft PR on an nRF5340 Audio DK, stressing it as much as I could today (connect/disconnect/restarting source) and it has not failed yet. Audio was crisp and clear when reconnecting to the source.

Thalley commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -43,4 +43,28 @@ config ENABLE_LC3
+	  Instead of using a broadcast assistant, sync automatically to first broadcast which
+	  has the configured TARGET_BROADCAST_CHANNEL.
+
+config TARGET_BROADCAST_CHANNEL
+	int "Broadcast Channel Audio Location to sync to"
+	# These corresponds to the bt_audio_location, supporting mono, left and right channels
+	range 0 2
+	default 1
+	help
+	  Channel Audio Location to sync to.
```
This should depend on `USE_USB_AUDIO_OUTPUT` - This makes the sample only able to sync to a single channel, but that's only an issue for the USB stuff, right?

Thalley commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -43,4 +43,28 @@ config ENABLE_LC3
+	depends on ENABLE_LC3
+	select USB_DEVICE_STACK
+	select USB_DEVICE_AUDIO
+	select RING_BUFFER
+
```
```suggestion
	select RING_BUFFER
	help
	  Enables USB audio as output as a USB peripheral. This does not support providing USB audio to e.g. speakers that are also USB peripherals, but can be connected to e.g. a phone or PC as a USB-in device (such as a USB microphone). 
	  USB audio only supports a single audio channel. 

```

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -83,99 +122,250 @@ static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+		/* For now we only handle one BIS, so always only decode the first element in
+		 * streams.
+		 */
+		do_lc3_decode(&streams[0]);
+#if defined(CONFIG_USB_DEVICE_AUDIO)
+		int err = 0;
+		enum bt_audio_location channels;
+
+		err = bt_audio_codec_cfg_get_chan_allocation(streams[0].stream.codec_cfg,
+							     &channels);
+
+		/* If the ring buffer usage is alrger than zero, then there is data to process */
+		if (ring_buf_space_get(&streams[0].audio_ring_buf)) {
+			mix_mono_to_stereo(channels);
+		}
+#endif /* #if defined(CONFIG_USB_DEVICE_AUDIO) */
```
While I understand that decoding multiple streams for USB is an issue, this also effectively modify the non-USB configuration to only support a single stream, compared to the current version of the sample that supports decoding multiple stream. 

I do not think the non-USB configuration should reduce its capabilities.

fredrikdanebjer replied:
I'm not entirely sure what you are referring to here. The code snippet you are highlighting here only runs for USB. What runs above is what ran prior to this PR. 
Assuming the Kconfigs you mentioned above are dependent on USB also, then there can also not be any modications to hos it syncs to the stream.

Thalley replied:
Sorry, it may not have been very clear:

This in a generic function that runs for both USB and non-USB builds, and you ever only do `do_lc3_decode(&streams[0]);` so this effectively reduces the sample's capabilities to decode all streams to just the first stream

fredrikdanebjer replied:
Aah, yes, thats correct, I will fix this

Thalley reviewed (changes requested):
The sample is still in some weird limbo state between support multiple streams only a single stream. 

The existing parts that supports multiple streams should not be reduced to a single stream, simply because we add USB support.

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -252,33 +472,55 @@ static struct bt_bap_stream_ops stream_ops = {
-	}
+	err = bt_bap_base_subgroup_bis_codec_to_codec_cfg(bis, &codec_cfg);
+
+	bt_audio_codec_cfg_get_chan_allocation(&codec_cfg, &chan_allocation);
```
Missing return check

Thalley commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -43,4 +43,35 @@ config ENABLE_LC3
+	  Instead of using a broadcast assistant, sync automatically to first broadcast which
+	  has the configured TARGET_BROADCAST_CHANNEL.
+
+config TARGET_BROADCAST_CHANNEL
+	int "Broadcast Channel Audio Location to sync to"
+	range 0 2
+	default 1
+	depends on USE_USB_AUDIO_OUTPUT
+	help
+	  Channel Audio Location to sync to. These corresponds to the bt_audio_location,
+	  supporting mono, left and right channels
```
Should this be optional? i.e. if not set, sync to the first BIS?

fredrikdanebjer replied:
I think currently this can make sense. We are acting as either left or right side (or mono). If the right channel does not exist, then there isn´t a stream we are interested in.

I understand where you come from, but I would prefer that we add this in an upcoming PR, along with possible stereo support.

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -252,33 +472,55 @@ static struct bt_bap_stream_ops stream_ops = {
+static int base_get_first_valid_bis(const struct bt_bap_base *base, uint32_t *bis_index)
+{
+	int err;
+	uint8_t valid_bis_index;
+
+
+	err = bt_bap_base_foreach_subgroup(base, find_valid_bis_in_subgroup_cb, &valid_bis_index);
```
```suggestion
	uint8_t valid_bis_index;

	err = bt_bap_base_foreach_subgroup(base, find_valid_bis_in_subgroup_cb, &valid_bis_index);
```

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -252,33 +472,55 @@ static struct bt_bap_stream_ops stream_ops = {
+	err = bt_bap_base_foreach_subgroup(base, find_valid_bis_in_subgroup_cb, &valid_bis_index);
+	if (err != -ECANCELED) {
+		printk("Failed to parse subgroups: %d\n", err);
+		return -1;
```
Suggest to return `errno` values or make the function a `bool`

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -252,33 +472,55 @@ static struct bt_bap_stream_ops stream_ops = {
+static int base_get_first_valid_bis(const struct bt_bap_base *base, uint32_t *bis_index)
+{
+	int err;
+	uint8_t valid_bis_index;
```
```suggestion
	uint8_t valid_bis_index = 0U;
```

If none are found then `bt_bap_base_foreach_subgroup` will still return 0 and then `valid_bis_index` will be uninitialized when you do `*bis_index |= ((uint8_t)1 << valid_bis_index);`

fredrikdanebjer replied:
If noone are found then the function will return beforehand, since `bt_bap_base_foreach_subgroup` will return 0 and not `-ECANCELED`.
I'm adding the initialization regardless

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -293,31 +535,19 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
-	if (err < 0) {
-		printk("Error: cannot enable LC3 codec: %d", err);
+#if defined(CONFIG_TARGET_BROADCAST_CHANNEL) && defined(CONFIG_USE_AUTOMATIC_BROADCAST_SYNC)
+	err = base_get_first_valid_bis(base, &base_bis_index_bitfield);
```
This function may return `0` where `base_bis_index_bitfield` gets set to 0, which isn't handled

fredrikdanebjer replied:
I dont think I see how that happens, can you elaborate?

`base_get_first_valid_bis` only returns 0 if the foreach returned `-ECANCELED`. Even if the bis index was zero, the return is a `1` bitshifted by the index - which would become ´1´ in that case.

What am I missing? :)

Thalley replied:
As I read it, it works as following

`base_get_first_valid_bis` returns 0 if `bt_bap_base_foreach_subgroup` returns 0. 
`bt_bap_base_foreach_subgroup` returns 0 if `bt_bap_base_subgroup_foreach_bis` returns anything but `-ECANCELED` (including 0). 
`bt_bap_base_subgroup_foreach_bis` returns 0 if channel allocation is not found. 

(Please correct me if the above is not correct)
So that means that if channel allocation is not found in the BIS codec configuration, then `base_get_first_valid_bis` returns 0 and the BIS index is not set, which means that this function can return 0, and `base_bis_index_bitfield` remain at `0`. In this case, `sem_bis_sync_requested` may still be set, even though `base_bis_index_bitfield` is 0, which is not correct. 

Summary: If the channel allocation is not set in any of the BIS (which is still a valid configuration/BASE), then `base_bis_index_bitfield` ends up being `0`, and then we give the `sem_bis_sync_requested` which is incorrect.

fredrikdanebjer replied:
>base_get_first_valid_bis returns 0 if bt_bap_base_foreach_subgroup returns 0.
This is incorrect :)
`base_get_first_valid_bis` returns 0 only if `bt_bap_base_foreach_subgroup` returns `-ECANCELED`. At time of your original comment it returned ´-1´ when `bt_bap_base_foreach_subgroup` returned 0, but has since been changed to return `-ENOENT`.

fredrikdanebjer commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -43,4 +43,35 @@ config ENABLE_LC3
+	  phone or PC as a USB-in device (such as a USB microphone).
+	  USB audio only supports a single audio channel.
+
+config USE_AUTOMATIC_BROADCAST_SYNC
+	bool "Automatic Broadcast Sync"
+	default y
+	depends on USE_USB_AUDIO_OUTPUT
+	help
+	  Instead of using a broadcast assistant, sync automatically to first broadcast which
+	  has the configured TARGET_BROADCAST_CHANNEL.
```
From the start it was intended that without this the Broadcast assistant will still be able to decide which BIS it uses, assuming it has the configured channel. However, its pointless with the current implementation, and I will remove it.

Thalley reviewed (changes requested):
One minor comment left, otherwise LGTM
